source/s3_platform_info.c

```diff
@@ -0,0 +1,592 @@
+    struct imds_callback_info *info = arg;
+    return info->shutdown_completed;
+}
+
+static void s_imds_client_on_get_instance_info_callback(
+    const struct aws_imds_instance_info *instance_info,
+    int error_code,
+    void *user_data) {
+    struct imds_callback_info *info = user_data;
+
+    aws_mutex_lock(&info->mutex);
+    if (error_code) {
+        info->error_code = error_code;
+    } else {
+        info->instance_type = aws_string_new_from_cursor(info->allocator, &instance_info->instance_type);
+    }
+    aws_condition_variable_notify_all(&info->c_var);
+    aws_mutex_unlock(&info->mutex);
+}
+
+static bool s_completion_predicate(void *arg) {
+    struct imds_callback_info *info = arg;
+    return info->error_code != 0 || info->instance_type != NULL;
+}
+
+struct aws_string *s_query_imds_for_instance_type(struct aws_allocator *allocator) {
+
+    struct imds_callback_info callback_info = {
+        .mutex = AWS_MUTEX_INIT,
+        .c_var = AWS_CONDITION_VARIABLE_INIT,
+        .allocator = allocator,
+    };
+
+    struct aws_event_loop_group *el_group = NULL;
+    struct aws_host_resolver *resolver = NULL;
+    struct aws_client_bootstrap *client_bootstrap = NULL;
+    /* now call IMDS */
+    el_group = aws_event_loop_group_new_default(allocator, 1, NULL);
+
+    if (!el_group) {
+        goto tear_down;
+    }
+
+    struct aws_host_resolver_default_options resolver_options = {
+        .max_entries = 1,
+        .el_group = el_group,
+    };
+
+    resolver = aws_host_resolver_new_default(allocator, &resolver_options);
+
+    if (!resolver) {
+        goto tear_down;
+    }
+
+    struct aws_client_bootstrap_options bootstrap_options = {
+        .event_loop_group = el_group,
+        .host_resolver = resolver,
+    };
+
+    client_bootstrap = aws_client_bootstrap_new(allocator, &bootstrap_options);
+
+    if (!client_bootstrap) {
+        goto tear_down;
+    }
+
+    struct aws_imds_client_shutdown_options imds_shutdown_options = {
+        .shutdown_callback = s_imds_client_shutdown_completed,
+        .shutdown_user_data = &callback_info,
+    };
+
+    struct aws_imds_client_options imds_options = {
+        .bootstrap = client_bootstrap,
+        .imds_version = IMDS_PROTOCOL_V2,
+        .shutdown_options = imds_shutdown_options,
+    };
+
+    struct aws_imds_client *imds_client = aws_imds_client_new(allocator, &imds_options);
+
+    if (!imds_client) {
+        goto tear_down;
+    }
+
+    aws_mutex_lock(&callback_info.mutex);
+
+    aws_imds_client_get_instance_info(imds_client, s_imds_client_on_get_instance_info_callback, &callback_info);
+    aws_condition_variable_wait_for_pred(
+        &callback_info.c_var, &callback_info.mutex, AWS_TIMESTAMP_SECS, s_completion_predicate, &callback_info);
+
+    aws_imds_client_release(imds_client);
+    aws_condition_variable_wait_pred(
+        &callback_info.c_var, &callback_info.mutex, s_client_shutdown_predicate, &callback_info);
+
```

Comment:
need to check return value of aws_imds_client_get_instance_info()

if it fails, don't wait for the result

but you'll still need to release and wait for shutdown